# Taller 2: Procesamiento de Lenguaje Natural
## Identificación Automática de Idiomas usando Redes LSTM

**Autores:**
- Andrés Rodriguez
- David Orozco  
- Javier Yela
- Alberto Torres

**Institución:** Universidad ICESI  
**Curso:** Procesamiento de Lenguaje Natural  
**Fecha:** 23 de Agosto de 2025  


---

## Resumen Ejecutivo

Este proyecto implementa un sistema de identificación automática de idiomas utilizando redes neuronales LSTM (Long Short-Term Memory). Se trabajó con el Language Identification Dataset, que contiene 90,000 ejemplos de fragmentos de texto en 20 idiomas diferentes. El modelo desarrollado alcanzó una precisión del 94.01% en el conjunto de prueba, demostrando su efectividad para la clasificación multilingüe.

---

## 1. Introducción

### 1.1 Contexto del Problema

La identificación automática de idiomas es una tarea fundamental en el procesamiento de lenguaje natural que consiste en determinar el idioma en el que está escrito un texto dado. Esta capacidad es esencial para:

- Sistemas de traducción automática
- Análisis de contenido en redes sociales
- Clasificación de documentos multilingües
- Preprocesamiento de datos textuales

### 1.2 Objetivo

El objetivo principal de este proyecto es desarrollar y evaluar un modelo de clasificación multiclase basado en LSTM para identificar automáticamente el idioma de fragmentos de texto entre 20 idiomas diferentes.

### 1.3 Dataset

El dataset utilizado proviene de la plataforma Hugging Face y contiene datos recolectados de tres fuentes principales:

1. **Multilingual Amazon Reviews Corpus** - Reseñas de productos en múltiples idiomas
2. **XNLI (Cross-lingual Natural Language Inference)** - Datos de inferencia textual multilingüe  
3. **STSb Multi MT (Semantic Textual Similarity Benchmark Multilingual Machine Translation)** - Datos de similitud semántica multilingüe

#### 1.3.1 Características del Dataset

- **Tamaño total:** 90,000 ejemplos
- **Número de idiomas:** 20
- **Idiomas incluidos:** Árabe (ar), Búlgaro (bg), Alemán (de), Griego moderno (el), Inglés (en), Español (es), Francés (fr), Hindi (hi), Italiano (it), Japonés (ja), Holandés (nl), Polaco (pl), Portugués (pt), Ruso (ru), Swahili (sw), Tailandés (th), Turco (tr), Urdu (ur), Vietnamita (vi), y Chino (zh)

#### 1.3.2 Estructura de los Datos

El dataset está compuesto por dos campos principales:
- **labels:** Etiqueta de texto que indica el idioma (ejemplo: "fr")
- **text:** Fragmento de texto en el idioma correspondiente (ejemplo: "Conforme à la description, produit pratique.")

### 1.4 Referencias

- Dataset: https://huggingface.co/datasets/papluca/language-identification
- Modelo de referencia: papluca/xlm-roberta-base-language-detection (99.6% de exactitud)

In [2]:
import pkg_resources
import warnings

warnings.filterwarnings('ignore')

installed_packages = [package.key for package in pkg_resources.working_set]
IN_COLAB = 'google-colab' in installed_packages

/tmp/ipython-input-2396000874.py:1: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  import pkg_resources


In [3]:
#!test '{IN_COLAB}' = 'True' && wget  https://github.com/Ohtar10/icesi-nlp/raw/refs/heads/main/requirements.txt && pip install -r requirements.txt
!test '{IN_COLAB}' = 'True' && sudo apt-get update -y
!test '{IN_COLAB}' = 'True' && sudo apt-get install python3.10 python3.10-distutils python3.10-lib2to3 -y
!test '{IN_COLAB}' = 'True' && sudo update-alternatives --install /usr/local/bin/python python /usr/bin/python3.11 2
!test '{IN_COLAB}' = 'True' && sudo update-alternatives --install /usr/local/bin/python python /usr/bin/python3.10 1
!test '{IN_COLAB}' = 'True' && pip install lightning datasets

Get:1 https://cli.github.com/packages stable InRelease [3,917 B]
Get:2 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]
Hit:3 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease
Get:4 https://cli.github.com/packages stable/main amd64 Packages [346 B]
Get:5 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]
Hit:6 http://archive.ubuntu.com/ubuntu jammy InRelease
Get:7 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]
Get:8 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]
Hit:9 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease
Hit:10 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease
Get:11 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages [9,190 kB]
Hit:12 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease
Get:13 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]
Get:14 http://security.

### Cargando el dataset
El dataset contiene 90.000 ejemplos de fragmentos de texto junto con su etiqueta de idioma, está disponible desde hugging face haciendo uso de la libreria datasets

In [4]:
from datasets import load_dataset
import warnings
import os

warnings.filterwarnings("ignore")
os.environ['TOKENIZERS_PARALLELISM'] = 'false'
dataset = load_dataset('papluca/language-identification', split='train')
dataset

README.md: 0.00B [00:00, ?B/s]

train.csv:   0%|          | 0.00/12.0M [00:00<?, ?B/s]

valid.csv: 0.00B [00:00, ?B/s]

test.csv: 0.00B [00:00, ?B/s]

Generating train split:   0%|          | 0/70000 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/10000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/10000 [00:00<?, ? examples/s]

Dataset({
    features: ['labels', 'text'],
    num_rows: 70000
})

El dataset contiene dos columnas "labels" que es la identificacion del idioma y "text" texto del mensaje, los idiomas que contiene el DS son 20 (arabic (ar), bulgarian (bg), german (de), modern greek (el), english (en), spanish (es), french (fr), hindi (hi), italian (it), japanese (ja), dutch (nl), polish (pl), portuguese (pt), russian (ru), swahili (sw), thai (th), turkish (tr), urdu (ur), vietnamese (vi), and chinese (zh))

In [5]:
dataset[1]

{'labels': 'bg',
 'text': 'размерът на хоризонталната мрежа може да бъде по реда на няколко километра ( km ) за на симулация до около 100 km за на симулация .'}

Con el siguiente código se obtiene una primera idea de la distribución de tamaños de los textos, lo cual es importante en tareas de procesamiento de lenguaje natural.

Conocer estas medidas permite:

Identificar textos inusualmente cortos o largos que podrían afectar el entrenamiento.

Decidir estrategias de preprocesamiento como truncar o rellenar (padding) los textos para que tengan una longitud uniforme.

Entender mejor las características del dataset antes de aplicar un modelo de clasificación.

In [6]:
text_lengths = [len(row['text']) for row in dataset]
print(f"Texto más corto: {min(text_lengths)}")
print(f"Texto más largo: {max(text_lengths)}")
print(f"Longitud promedio: {sum(text_lengths) / len(text_lengths)}")

Texto más corto: 2
Texto más largo: 2422
Longitud promedio: 110.86141428571429


## Definiendo el Tokenizer

## Tokenizador (simple_tokenizer)

Convierte todo el texto a minúsculas.

Separa palabras por espacios.

Conservar caracteres especiales es útil, ya que el dataset incluye múltiples lenguas con acentos, caracteres propios (ñ, ü, ç, etc.), e incluso alfabetos distintos (árabe, chino).

### Tokens: 30.000

Un vocabulario amplio permite capturar la mayoría de palabras frecuentes en los 20 idiomas del dataset.

Si fuera muy pequeño, se perdería contexto porque muchas palabras se reemplazarían por [UNK].

Si fuera demasiado grande, aumentaría innecesariamente la memoria y tiempo de entrenamiento.

30k es un punto de equilibrio: suficiente para capturar diversidad lingüística sin sobredimensionar el modelo.

### Max_length 100:

La longitud promedio de los textos era de ~111 caracteres (no tokens), y muchos mensajes son cortos (ej. frases de 5–20 palabras).

Usar 100 tokens como máximo permite:

Capturar suficiente contexto en la mayoría de casos.

Mantener la eficiencia en cómputo (no desperdiciar memoria con secuencias muy largas que son raras).

Este valor es ideal porque combina la cobertura de contexto con la eficiencia del modelo.

In [7]:
import re
from collections import Counter

def simple_tokenizer(text):
    text = text.lower()
    return text.strip().split()

# Construimos el vocabulario a partir de conjunto de datos.
token_counts = Counter()
for text in dataset["text"]:
    token_counts.update(simple_tokenizer(text))

# 50k-2 porque necesitamos reservar espacio para los dos tokens especiales
top_n_tokens = list(token_counts.keys())[:30000-2]
vocab = {"[PAD]": 0, "[UNK]": 1}
for token in top_n_tokens:
    vocab[token] = len(vocab)

def tokenize_text(text, max_length=100):
    tokens = text.lower().split()
    ids = [vocab.get(tok, vocab["[UNK]"]) for tok in tokens[:max_length]]
    ids += [vocab["[PAD]"]] * (max_length - len(ids))
    return ids

Exploracion de tokens, así verificamos completitud de las palabras<.>

In [8]:
print(f"Vocabulario: {len(vocab)} tokens")
print("Primeros 15 tokens:")
print(f"{top_n_tokens[:15]}")
print("15 tokens de en medio:")
print(f"{top_n_tokens[1000:1015]}")
print("Últimos 15 tokens:")
print(f"{top_n_tokens[-15:]}")

Vocabulario: 30000 tokens
Primeros 15 tokens:
['os', 'chefes', 'de', 'defesa', 'da', 'estónia,', 'letónia,', 'lituânia,', 'alemanha,', 'itália,', 'espanha', 'e', 'eslováquia', 'assinarão', 'o']
15 tokens de en medio:
['bằng', 'tiếng', 'anh', 'sirve', 'nada,', 'mala', 'adherencia,', 'material', 'muy', 'debil', 'демократична', 'конвенция', 'обаче', 'остави', 'макгавърн']
Últimos 15 tokens:
['section', 'postpartum', 'recovery', 'period.', 'panty', 'incision', 'scratches', 'incision.', 'sorts', 'methods,', 'lowering', 'waist,', 'wearing', 'loose,', 'etc.']


In [9]:
tokenized = tokenize_text("hola du sun nuit bon", max_length=8)
tokenized

[3647, 538, 19341, 13869, 624, 0, 0, 0]

Cada palabra fue convertida en un número entero (ID de token) según el vocabulario:

"hola" → 3647

"du" → 538

"sun" → 19341

"nuit" → 13869

"bon" → 624

Los valores corresponden a la posición de cada palabra en el vocabulario construido con los 30.000 tokens más frecuentes.

In [10]:
id_2_token = {v: k for k, v in vocab.items()}
[id_2_token[token] for token in tokenized]

['hola', 'du', 'sun', 'nuit', 'bon', '[PAD]', '[PAD]', '[PAD]']

Sólo como ejemplo, y a modo de prueba, pudimos capturar en varios idiomas palabras aleatorias al revisar que con un número reducido de tokens, algunas palabras no se capturan y un número más alto aumentaba el tiempo de procesamiento pero no influía mucho en el rendimiento.

### Definiendo el dataset de pytorch
Convierte el dataset de texto crudo en pares (tokens, etiqueta numérica) listos para entrenar un modelo de clasificación de idiomas en PyTorch.

El parámetro seq_length define la longitud máxima de tokens que tendrá cada texto al ser tokenizado.

Si un texto es más corto que 128 tokens, se completa con [PAD].

Si un texto es más largo, se trunca a ese tamaño.

seq_length=128 se selecciona como valor óptimo porque balancea cobertura de información y eficiencia en entrenamiento, evitando tanto pérdida de contexto como exceso de cómputo.

In [11]:
import torch
import numpy as np
from typing import Tuple, Dict
from torch.utils.data import Dataset

class LanguageIdentification(Dataset):

    def __init__(self, tokenizer, dataset, seq_length: int = 128):
        self.tokenizer = tokenizer
        self.dataset = dataset
        self.seq_length = seq_length
        self.id_2_class_map = dict(enumerate(np.unique(dataset[:]['labels'])))
        self.class_2_id_map = {v: k for k, v in self.id_2_class_map.items()}
        self.num_classes = len(self.id_2_class_map)

    def __getitem__(self, index) -> Dict[str, torch.Tensor]:
        text, y = self.dataset[index]['text'], self.dataset[index]['labels']
        y = self.class_2_id_map[y]
        data = {'input_ids': torch.tensor(self.tokenizer(text, max_length=self.seq_length))}
        data['y'] = torch.tensor(y)
        return data


    def __len__(self):
        return len(self.dataset)

Este bloque prepara el dataset final para ser usado en el entrenamiento, asegurando que todos los ejemplos estén correctamente tokenizados y listos para entrar a un DataLoader

In [12]:
languageIdentification_dataset = LanguageIdentification(tokenize_text, dataset)
assert len(languageIdentification_dataset) == len(dataset)

Y luego, procedemos a hacer el train-val-test split y crear los dataloaders.

In [13]:
from torch.utils.data import random_split
from torch.utils.data import DataLoader

batch_size = 4 if not IN_COLAB else 12
train_dataset, val_dataset, test_dataset = random_split(languageIdentification_dataset, lengths=[0.8, 0.1, 0.1])
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True, num_workers=2)
val_loader = DataLoader(val_dataset, batch_size=batch_size, shuffle=False, num_workers=2)
test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False, num_workers=2)

## Definición del modelo LSTM

1. El bloque recibe una secuencia de tokens

2. Convierte los tokens en vectores mediante la capa de embeddings.

3. Procesa la secuencia con una red LSTM que aprende dependencias y relaciones entre palabras.

4. Devuelve el último estado oculto de la LSTM, que funciona como un resumen de todo el texto y sirve de entrada a una capa de clasificación posterior.

In [14]:
import torch
import torch.nn as nn
import torch.nn.functional as F

class LSTMBlock(nn.Module):
    def __init__(self, vocab_size, embed_dim, hidden_dim, num_classes, num_layers=2, dropout=0.2):
        super().__init__()
        self.embedding = nn.Embedding(vocab_size, embed_dim, padding_idx=0)
        self.lstm = nn.LSTM(embed_dim, hidden_dim, num_layers, batch_first=True, dropout=dropout)

    def forward(self, x):
        embedded = self.embedding(x)
        output, (hidden, _) = self.lstm(embedded)
        return hidden[-1]


### Definición del clasificador

Finalmente, definimos el modelo en sí. Este modelo constará de 3 capas:

- La tokenización, tal como la definimos anteriormente.
- El bloque LSTM, que acabamos de definir.
- Una capa densa adicional que servirá como clasificador de aquello que nos entregue la capa del transformer.

Como este es un LightningModule, aquí definiremos el resto de funciones utilitarias para el entrenamiento de la tarea.

In [15]:
from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.loggers import TensorBoardLogger
from pytorch_lightning.callbacks.early_stopping import EarlyStopping
from torchmetrics import Accuracy

class LanguageClassifierWithLSTM(LightningModule):

    def __init__(self, vocab_size: int, num_classes: int, emb_dim: int, hidden_dim: int = 128):
        super(LanguageClassifierWithLSTM, self).__init__()
        self.num_classes = num_classes
        self.lstm = LSTMBlock(vocab_size, emb_dim, hidden_dim, num_classes)
        self.classifier = nn.Sequential(
            nn.Flatten(),
            nn.Linear(hidden_dim, 512),
            nn.ReLU(),
            nn.Dropout(0.2),
            nn.Linear(512, 256),
            nn.ReLU(),
            nn.Dropout(0.2),
            nn.Linear(256, num_classes),
            nn.LogSoftmax(dim=1)
        )

        self.train_acc = Accuracy(task='multiclass', num_classes=num_classes)
        self.val_acc = Accuracy(task='multiclass', num_classes=num_classes)
        self.test_acc = Accuracy(task='multiclass', num_classes=num_classes)

    def forward(self, x):
        embeddings = self.lstm(x)
        return self.classifier(embeddings)


    def training_step(self, batch, batch_idx):
        x, y = batch['input_ids'], batch['y']
        # print(f"\nbatch-idx: {batch_idx}")
        # print(f"shape of x: {x.shape}")
        # print(torch.max(x, dim=0))
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y)
        self.train_acc(y_hat, y)
        self.log('train-loss', loss, prog_bar=True, on_step=False, on_epoch=True)
        self.log('train-acc', self.train_acc, prog_bar=True, on_step=False, on_epoch=True)
        return loss

    def validation_step(self, batch):
        x, y = batch['input_ids'], batch['y']
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y)
        self.val_acc(y_hat, y)
        self.log('val-loss', loss, prog_bar=True, on_step=False, on_epoch=True)
        self.log('val-acc', self.val_acc, prog_bar=True, on_step=False, on_epoch=True)
        return loss

    def test_step(self, batch):
        x, y = batch['input_ids'], batch['y']
        y_hat = self(x)
        self.test_acc(y_hat, y)
        self.log('test-acc', self.test_acc, prog_bar=True, on_step=False, on_epoch=True)


    def predict_step(self, batch):
        x = batch['input_ids']
        return self(x)


    def configure_optimizers(self):
        optimizer =  torch.optim.AdamW(self.parameters(), lr=1e-3, weight_decay=1e-5)
        return optimizer


model = LanguageClassifierWithLSTM(vocab_size=len(vocab) + 1, num_classes=languageIdentification_dataset.num_classes, emb_dim=512)

tb_logger = TensorBoardLogger('tb_logs', name='LSTMClassifier')
callbacks=[EarlyStopping(monitor='train-loss', patience=3, mode='min')]
trainer = Trainer(max_epochs=10, devices=1, logger=tb_logger, callbacks=callbacks, precision="16-mixed")

trainer.fit(model, train_dataloaders=train_loader, val_dataloaders=val_loader)

INFO:pytorch_lightning.utilities.rank_zero:Using 16bit Automatic Mixed Precision (AMP)
INFO:pytorch_lightning.utilities.rank_zero:💡 Tip: For seamless cloud uploads and versioning, try installing [litmodels](https://pypi.org/project/litmodels/) to enable LitModelCheckpoint, which syncs automatically with the Lightning model registry.
INFO:pytorch_lightning.utilities.rank_zero:GPU available: True (cuda), used: True
INFO:pytorch_lightning.utilities.rank_zero:TPU available: False, using: 0 TPU cores
INFO:pytorch_lightning.utilities.rank_zero:HPU available: False, using: 0 HPUs
INFO:pytorch_lightning.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
INFO:pytorch_lightning.callbacks.model_summary:
  | Name       | Type               | Params | Mode 
----------------------------------------------------------
0 | lstm       | LSTMBlock          | 15.8 M | train
1 | classifier | Sequential         | 202 K  | train
2 | train_acc  | MulticlassAccuracy | 0      | train
3 | val_acc    | M

Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

INFO:pytorch_lightning.utilities.rank_zero:`Trainer.fit` stopped: `max_epochs=10` reached.


Observemos el proceso de entrenamiento

In [16]:
%load_ext tensorboard

In [17]:
%tensorboard --logdir tb_logs/

<IPython.core.display.Javascript object>

Tras entrenar el modelo, se realizó la evaluación en el conjunto de prueba utilizando trainer.test.

Métrica reportada: test-acc

Valor obtenido: ≈ 0.94 (94%)

Interpretación

El modelo logró clasificar correctamente el idioma del texto en más del 94% de los casos.

Este resultado es alto considerando que la tarea abarca 20 idiomas diferentes, algunos de ellos con similitudes léxicas (ej. español, italiano, portugués) que pueden inducir a confusión.

El desempeño confirma que la arquitectura LSTM implementada, junto con el preprocesamiento (vocabulario de 30k tokens, longitud de secuencia fija, embeddings, etc.), fue suficiente para aprender representaciones robustas para la clasificación multilingüe.

In [18]:
model.eval()
trainer.test(model, test_loader)

INFO:pytorch_lightning.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]


Testing: |          | 0/? [00:00<?, ?it/s]

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│         test-acc          │    0.9417142868041992     │
└───────────────────────────┴───────────────────────────┘

[{'test-acc': 0.9417142868041992}]

### Haciendo predicciones

Finalmente, vamos a hacer uso del modelo y ver que tan bueno es para la clasificación de lenguaje.

In [19]:
predictions = trainer.predict(model, test_loader)
predictions = torch.cat(predictions, dim=0)
predictions = torch.argmax(predictions, dim=-1)
predictions = [languageIdentification_dataset.id_2_class_map[pred] for pred in predictions.numpy()]

INFO:pytorch_lightning.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]


Predicting: |          | 0/? [00:00<?, ?it/s]

Verificacion de clasificacion del modelo

In [20]:
import pandas as pd

test_indices = test_dataset.indices
df = pd.DataFrame(data={
    "texto": dataset[test_indices]['text'],
    "tokens": [tokenize_text(v) for v in dataset[test_indices]['text']],
    "labels": dataset[test_indices]['labels'],
    'predicción': predictions
}, index=test_indices)

df['tokens_string'] = df.tokens.apply(lambda t: ' '.join([id_2_token[i] for i in t]))
df = df[["texto", "tokens", "tokens_string", "labels", "predicción"]]
df.style.set_table_styles(
    [
        {'selector': 'td', 'props': [('word-wrap', 'break-word')]}
    ]
)
df.head(15)

texto  \
31265  Mir gefällt die Art, wie der Autor die Handlun...   
65208  I purchased this headset for the over the ear ...   
15519  Chương trình đã cung cấp tác phẩm , bất chấp n...   
55545  नीचे पड ़ ोसी हारून whiteheard ने कहा , एक रात...   
28583                     Leider sehr ungenaue ohm werte   
18636           स ् वर ् ग से पृथ ् वी पर जोर दिया गया ।   
31162        包包收到了很喜欢。质量很好，做工精细，携带很方便皮质很柔软摸起来手感也很不错，很满意！   
50214                     No cubre todo, se queda corto.   
5715   سام ، ایک بارگی سام ' گنڈالف ' کی جانب سے ایک ...   
57913  iran insiste que o programa se destina exclusi...   
51324  มัน เป็น ชายหนุ่ม ที่ กระวนกระวายใจ ของ เรา ขอ...   
57594  bonjour comment séparer la lampe et le capteur...   
37823           Delivered on time and material is great!   
40917  O homem está sentado a uma mesa desarrumada e ...   
42264               棉的挺舒服的..............................   

                                                  tokens  \
31265  [289, 4688, 244, 1, 265, 282, 1, 244, 1, 282, ...   
65208  [1196, 12891, 827, 18676, 806, 822, 1771, 822,...   
15519  [4386, 12434, 866, 5816, 4230, 22046, 8267, 98...   
55545  [17484, 7479, 1830, 1, 1, 1, 212, 904, 98, 221...   
28583  [557, 778, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   
18636  [3388, 208, 6325, 208, 9026, 224, 1, 208, 1782...   
31162  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
50214  [390, 13200, 21206, 409, 6799, 1, 0, 0, 0, 0, ...   
5715   [1, 76, 152, 1, 1, 137, 1, 137, 79, 1183, 151,...   
57913  [7788, 1, 174, 16, 1, 409, 11597, 1, 841, 1, 1...   
51324  [655, 469, 1, 46, 1, 474, 6591, 474, 1, 0, 0, ...   
57594  [26872, 2111, 1, 180, 6551, 540, 499, 1, 551, ...   
37823  [8349, 1731, 1766, 1192, 1009, 1198, 3256, 0, ...   
40917  [16, 984, 2309, 4953, 841, 988, 8845, 1, 13, 8...   
42264  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   

                                           tokens_string labels predicción  
31265  mir gefällt die [UNK] wie der [UNK] die [UNK] ...     de         de  
65208  i purchased this headset for the over the ear ...     en         en  
15519  chương trình đã cung cấp tác phẩm , bất [UNK] ...     vi         vi  
55545  नीचे पड ़ [UNK] [UNK] [UNK] ने कहा , एक रात , ...     hi         hi  
28583  leider sehr [UNK] [UNK] [UNK] [PAD] [PAD] [PAD...     de         de  
18636  स ् वर ् ग से [UNK] ् वी पर [UNK] दिया गया । [...     hi         hi  
31162  [UNK] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD...     zh         zh  
50214  no cubre todo, se queda [UNK] [PAD] [PAD] [PAD...     es         es  
5715   [UNK] ، ایک [UNK] [UNK] ' [UNK] ' کی جانب سے ا...     ur         ur  
57913  iran [UNK] que o [UNK] se destina [UNK] a [UNK...     pt         pt  
51324  มัน เป็น [UNK] ที่ [UNK] ของ เรา ของ [UNK] [PA...     th         th  
57594  bonjour comment [UNK] la lampe et le [UNK] ???...     fr         fr  
37823  delivered on time and material is great! [PAD]...     en         en  
40917  o homem está sentado a uma mesa [UNK] e a usar...     pt         pt  
42264  [UNK] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD...     zh         zh

Podemos ver las "NO" coincidencias, donde el modelo no clasifica bien el idioma del texto

In [21]:
errors = df[df['labels'] != df['predicción']]
errors.head(15)

texto  \
42560                        A tartaruga seguiu o peixe.   
21216  暗証番号設定のダイヤルが回らない。もう一つ買っていたのでそちらも試したところそちらは上手く回...   
34813          運動後の体のケアに使っています。ゴツゴツしていて痛気持ちいいのが気に入っています。   
44754  3 extremistas suspeitos foram libertados sob f...   
43001         非常にクセがあります。飲みやすくはありません。好みが分かれるウイスキーだと思います。   
42906  はめたらすぐ亀裂が入って残念でした･･･手作りとはいえ、もぉちょっと縫い目などが綺麗だとよか...   
24168         商品に文句はありませんが、ショップと連絡がつきません。なのでもう買わないと思います。   
38930  前回購入したBluetooth5.0は仕事用で購入いたしましたが、今回購入した理由は、自宅内...   
34824                              autobus jadący ulicą.   
68096                                  Lazima act haraka   
39900  Tại tập đoàn dịch vụ tài chính , nhóm an ninh ...   
60754  第一次在亚马逊上面购买图书就遇到这样的问题，看来品控还不如淘宝上面任何一个店家强！整个系列书...   
32953                   Oh i ' ve done labda nne au tano   
2243   持ちやすいが、ホントに0.4なのと思う。手帳用に購入したが、0.5のジェットストリームより細...   
54957  友達の誕生日にネタで渡しました。感想をきいてみたんですが（笑）、リアルな質感で本物みたいだっ...   

                                                  tokens  \
42560  [841, 1, 1, 16, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
21216  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
34813  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
44754  [1364, 1, 1, 4090, 1, 1, 1, 0, 0, 0, 0, 0, 0, ...   
43001  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
42906  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
24168  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
38930  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
34824  [3428, 20459, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   
68096  [8830, 23624, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   
39900  [991, 2095, 1, 12027, 6125, 2273, 2099, 98, 1,...   
60754  [1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
32953  [1428, 1196, 137, 1072, 10597, 7370, 1, 599, 1...   
2243   [17936, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   
54957  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   

                                           tokens_string labels predicción  
42560  a [UNK] [UNK] o [UNK] [PAD] [PAD] [PAD] [PAD] ...     pt         pl  
21216  [UNK] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD...     ja         zh  
34813  [UNK] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD...     ja         zh  
44754  3 [UNK] [UNK] foram [UNK] [UNK] [UNK] [PAD] [P...     pt         nl  
43001  [UNK] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD...     ja         zh  
42906  [UNK] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD...     ja         zh  
24168  [UNK] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD...     ja         zh  
38930  [UNK] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD...     ja         zh  
34824  autobus jadący [UNK] [PAD] [PAD] [PAD] [PAD] [...     pl         ja  
68096  lazima act [UNK] [PAD] [PAD] [PAD] [PAD] [PAD]...     sw         zh  
39900  tại tập [UNK] dịch vụ tài chính , [UNK] an nin...     vi         ru  
60754  [UNK] [UNK] [UNK] [PAD] [PAD] [PAD] [PAD] [PAD...     zh         ja  
32953  oh i ' ve done labda [UNK] au [UNK] [PAD] [PAD...     sw         tr  
2243   持ちやすいが、ホントに0.4なのと思う。手帳用に購入したが、0.5のジェットストリームより細...     ja         zh  
54957  [UNK] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD...     ja         zh

El modelo LSTM alcanzó un desempeño sobresaliente en la tarea de identificación de idiomas, con una precisión global de ~94%.

Los idiomas de escritura única (árabe, chino, ruso, hindi, japonés) fueron clasificados con gran exactitud.

En los idiomas (español, francés, italiano, portugués), donde cabría esperar mayor confusión por la cercanía léxica, los resultados muestran muy pocas equivocaciones, lo que indica que el modelo aprendió a diferenciar eficazmente sus patrones.

La principal fuente de error aparece entre japonés y chino, una confusión razonable por compartir caracteres visualmente similares.

In [24]:
from sklearn.metrics import confusion_matrix
import pandas as pd

# Recuperar el mapeo desde el dataset instanciado
id_2_class_map = languageIdentification_dataset.id_2_class_map
class_2_id_map = languageIdentification_dataset.class_2_id_map

# Convertir predicciones al mismo espacio que las etiquetas reales
if isinstance(predictions[0], (int, np.integer)):
    y_pred = [id_2_class_map[p] for p in predictions]
else:
    y_pred = predictions

y_true = dataset[test_dataset.indices]['labels']

# Construir la matriz de confusión
labels = list(id_2_class_map.values())
cm = confusion_matrix(y_true, y_pred, labels=labels)

# Mostrar como DataFrame para legibilidad
cm_df = pd.DataFrame(cm, index=labels, columns=labels)
cm_df



ar   bg   de   el   en   es   fr   hi   it   ja   nl   pl   pt   ru   sw  \
ar  349    1    0    1    0    0    0    0    0    1    0    0    0    0    0   
bg    1  354    0    0    0    0    0    0    0    1    0    0    0    6    0   
de    0    0  346    0    0    0    1    0    0    0    0    0    0    0    1   
el    3    0    0  319    0    0    0    0    0    0    0    0    0    1    0   
en    0    0    0    0  356    0    0    0    1    1    0    1    0    0    0   
es    0    0    0    0    0  326    0    0    4    1    0    0    1    0    1   
fr    0    0    0    0    0    1  353    0    1    0    0    0    0    0    1   
hi    0    0    0    0    0    0    0  362    0    0    0    0    0    1    0   
it    1    0    0    0    0    0    0    0  334    2    1    1    0    0    0   
ja    1    1    0    0    0    0    0    1    0  183    0    1    0    2    0   
nl    0    0    0    0    2    0    0    0    2    7  332    3    0    0    0   
pl    0    0    0    0    0    0    0    0    1   12    5  321    0    0    0   
pt    0    0    0    0    0    1    0    0    3    1    9    7  339    0    0   
ru    2    5    0    0    0    0    0    0    0    0    0    0    0  304    0   
sw    1    0    0    0    4    0    0    0    1    0    0    1    0    0  325   
th    0    0    0    0    0    0    0    0    0    5    0    0    0    2    0   
tr    4    0    0    0    0    0    0    0    0    2    2    0    0    4    0   
ur    2    0    0    0    1    0    0    0    0    2    0    0    0    1    1   
vi    1    0    0    0    0    0    0    0    0    0    0    0    0    1    0   
zh    0    0    0    0    0    0    0    1    0   63    1    0    0    1    0   

     th   tr   ur   vi   zh  
ar    0    3    0    0    1  
bg    0    1    0    0    0  
de    0    0    0    0    2  
el    0    5    0    0    0  
en    0    0    0    0    1  
es    0    0    0    0    0  
fr    0    0    0    0    0  
hi    0    1    0    0    0  
it    0    1    0    0    0  
ja    2    1    2    0  165  
nl    0    2    0    0    0  
pl    0    1    0    0    1  
pt    0    1    0    0    0  
ru    1    0    0    0    0  
sw    0   10    0    0    2  
th  346    1    0    0    0  
tr    0  343    0    0    0  
ur    0    0  369    0    0  
vi    0    0    0  355    0  
zh    0    0    1    0  276

## Conclusiones

El trabajo desarrollado permitió construir un pipeline completo de identificación automática de idiomas utilizando una arquitectura basada en LSTM. A partir del preprocesamiento del dataset, la creación de un vocabulario de 30k tokens y la normalización de secuencias, se entrenó un modelo capaz de reconocer textos en 20 idiomas con una precisión superior al 94% en el conjunto de prueba. Los resultados de la matriz de confusión evidencian que el modelo clasifica con gran exactitud tanto idiomas con alfabetos únicos (árabe, chino, ruso, hindi, japonés) como lenguas (español, francés, italiano, portugués), en estas últimas con un nivel de error mínimo a pesar de su cercanía léxica. La principal confusión se presenta entre japonés y chino, un resultado esperable dada la similitud de caracteres en ambos sistemas de escritura.

En conclusión, el modelo demuestra ser eficiente, robusto y generalizable, confirmando la utilidad de las LSTM en tareas de clasificación multilingüe y estableciendo una base sólida para futuras extensiones con arquitecturas más avanzadas como Transformers.